README.md
## DCDencode: Encode DCD trajectory into Structural Alphabet sequences
[![release](https://img.shields.io/badge/release-v0.4-green?logo=github)](https://github.com/Fraternalilab/DCDencode)

The package provides the functionality to encode a given trajectory
into Structural Alphabet (SA) strings.


## Installation
There are several ways to install the *DCDencode* package, please choose one of the following.
You need to have R installed and the *bio3d* package.

### Linux shell
Download the [latest *DCDencode* release (.tar.gz)](https://github.com/Fraternalilab/DCDencode/releases/latest)
and install on the shell with (example for version 0.1):
```{sh}
R CMD INSTALL DCDencode-v.0.1.tar.gz
```

### R console
Download the [latest *DCDencode* release (.tar.gz)](https://github.com/Fraternalilab/DCDencode/releases/latest) and
install from the R console (example for version 0.1, assuming it is located in the current directory):
```{r}
install.packages("./DCDencode-v.0.1.tar.gz")
```

### R console with devtools
Install the *devtools* package and install *DCDencode* directly from GitHub:
```{r}
library("devtools")
install_github("Fraternalilab/DCDencode")
```


## Usage
### Encoding
Run the script *Rscripts/dcdencode.R* in the directory containing the DCD trajectory file:
```{sh}
Rscript pdbencode.R <structure_name>.pdb <trajectory_name>.dcd <confInc>
```
The first argument must be the reference PDB structure.
The second argument must be the DCD trajectory.
The third argument <confInc> is an optional conformation increment (integer)
that will be added to the number of the first conformation, which is always '1'.
The specified DCD trajectory will be encoded into a
'<structure\_name>.<trajectory\_name>.sasta' file
containing one SA sequence per trajectory conformation in FASTA-type format.
The format of SA sequence headers is \<ID\>|\<conformation\>,
where ID is a dot-separated structure and trajectory name,
while conformations are numbered from 1 to N (= total number of conformations),
plus the conformation increment (default = 0).

### Splitting DCD trajectories
Trajectories can be large and therefore the encoding routine can be memory hungry.
To avoid running out of RAM, a script employing the *catdcd* program (VMD suite)
is included, which facilitates splitting the trajectory into blocks:
```{sh}
Rscript dcdsplit.R <traj_name>.dcd <first> <last> <nConf>
```
After encoding all blocks, the resulting SA strings can be concatenated to the complete encoded trajectory.


#### Copyright Holders, Authors and Maintainers 
- 2021 Jens Kleinjung (author, maintainer) jens@jkleinj.eu
- 2021 Franca Fraternali (author, maintainer) franca.fraternali@kcl.ac.uk



NEWS
# Revision History of DCDencode Package 
- 0.3
Sun 28 Feb 13:36:57 GMT 2021
	Improvements in Rscript and progress bars

- 0.2
Mon 22 Feb 18:57:12 GMT 2021
	Package Rchecked and with man pages

- 0.1
Mon 15 Feb 19:09:55 GMT 2021
	Package revision started



R/kabsch.R
#===============================================================================
# DCDencode 
#' Kabsch Algorithm in R code
#'
#' Aligns two sets of points via rotation and translation after
#'   solving the superpositioning problem in an Eigenvalue Ansatz.
#'
#' Given two sets of points, with one specified as the reference set,
#' the other set will be rotated so that the RMSD between the two is minimized.
#' The format of the matrix is that there should be one row for each of
#' n observations, and the number of columns, d, specifies the dimensionality
#' of the points. The point sets must be of equal size and with the same
#' ordering, i.e. point one of the second matrix is mapped to point one of
#' the reference matrix, point two of the second matrix is mapped to point two
#' of the reference matrix, and so on.
#'
#' @param P n x d matrix of reference points.
#' @param Q n x d matrix of points to align to to \code{pm}
#' @return Matrix \code{qm} rotated and translated so that the ith point
#'	is aligned to the ith point of \code{pm} in the least-squares sense.
#' @references
#' \url{https://en.wikipedia.org/wiki/Kabsch_algorithm}
#'
#' (C) 2019 Jamie Macpherson and Jens Kleinjung
#===============================================================================

kabsch_R = function(Q, P) {
	## cast array into matrix
	dim(Q) = c(3, 4);
	Q.m = t(Q);

	dim(P) = c(3, 4);
	P.m = t(P);

	## center objects
	# center Q.m
	Qc = scale(Q.m, center = TRUE, scale = FALSE)

	# center P
	Pc = scale(P.m, center = TRUE, scale = FALSE)

	## determine the cross-covariance matrix
	C = crossprod(Pc, Qc)

	## compute a single value decomposition of the
	## cross-covariance matrix
	C.svd = svd(C)

	## use the sign of the determinant to ensure a right-hand coordinate system
	## (this is only required if the fragments are stereogenic systems, which
	## they aren't)
	#d = det(C.svd$u) * det(C.svd$v)

	## determine the optimal rotation matrix
	U = tcrossprod(C.svd$u, C.svd$v)

	## rotate matrix P unto Q
	Prot = Pc %*% U

	## determine the RMSD between rotated P and Q
	# squared differences
	Dsr.sq = (Prot - Qc)**2

	rmsd = sqrt(sum(apply(Dsr.sq, 1, sum)) / dim(Dsr.sq)[1]);

	## spit out the optimal fit RMSD between Q and P
	return(rmsd)
}

#===============================================================================



Rscripts/dcdsplit.R
#===============================================================================
# DCDencode
# dcdsplit.R
# Split trajectory into evenly sized files
# Requires 'catdcd' program (VMD software suite) 
# Usage: Rscript dcdsplit.R <traj> <first> <last> <nConf>
#===============================================================================

args = commandArgs(trailingOnly = TRUE)

## first argument is trajectory name
traj = args[1]
traj_name = unlist(strsplit(traj, "\\.", perl = TRUE))[1]

## second argument is first conformation, i.e. where to start splittig
first = abs(as.integer(args[2]))

## third argument is last conformation, i.e. where to end splittig
last = abs(as.integer(args[3]))

## fourth argument is target number of conformations in each output trajectory
nConf = abs(as.integer(args[4]))

error = tryCatch(stopifnot(nchar(args[1]) > 0, args[2] > 0, args[3] > 0, args[4] > 0),
  error = function(e) {
    ## stopifnot error message
    message(e)
    ## usage message
    message("\nUsage: 'Rscript dcdsplit.R <traj> <first> <last> <nConf>'")
    ## exit tryCatch with error code
    return(1)
  }
)
## stop if error code exists
if (! is.null(error)) { stop("Caught error in input arguments") }

#_______________________________________________________________________________
## number of output files with target number of conformations
nSplit = floor((last - first + 1) / nConf)
## additional output file for overhang conformations
isOverhang = as.numeric((last - first + 1) %% nConf != 0)

for (i in 0:(nSplit + isOverhang - 1)) {
  split_first = first + (i * nConf)
  ## last conformation must be larger than first
  stopifnot(last > split_first)
  split_last = min((split_first + nConf - 1), last)
  inname = sprintf("%s.dcd", traj_name)
  outname = sprintf("%s.%d-%d.dcd", traj_name, split_first, split_last)
  command = sprintf("catdcd -o %s -first %d -last %d %s",
                    outname, split_first, split_last, inname)
  message(command)
  system(command)
}

#===============================================================================


Rscripts/dcdencode.R
#===============================================================================
# DCDencode 
# dcdencode.R
# Read PDB structure and trajectory in local directory
# Encode trajectory and write SA strings as FASTA format
#===============================================================================

library("DCDencode")
library("bio3d")
library("pbapply")

#_______________________________________________________________________________
## by default no conformation increment (first conformation is '1 + confInc')
confInc = 0

## command line arguments
args = commandArgs(trailingOnly = TRUE)
## argument [1] is input structure
str = as.character(args[1])
## argument [2] is input trajectory
traj = as.character(args[2]) 
## argument [3] is conformation increment if trajectory has been split
confInc = abs(as.integer(args[3]))

error = tryCatch(stopifnot(nchar(args[1]) < 1, nchar(args[2]) < 1),
  error = function(e) {
    ## stopifnot error message
    message(e)
    ## usage message
    message("\nUsage: 'Rscript dcdsplit.R <str> <traj> <confInc>'")
    ## exit TryCatch with error code
    return(1)
  }
)
## stop if error code exists
if (! is.null(error)) { stop("Caught error in input arguments") }

#_______________________________________________________________________________
## args[1]: input structure
#args[1] = "/mnt/databases/MD/PKM2/3BJT_monomer_apo/md0/3BJT_md0_ca.pdb"
message("  The atom order and number of the input structure and input trajectory must match.\
Please note that potential mis-matches are unlikely to be detected by DCDencode.")
message(paste("Input structure:", str))
## structure name
str_name = unlist(strsplit(str, "\\.", perl = TRUE))[1]
## read structure
str_bio3d = bio3d::read.pdb2(str)

#_______________________________________________________________________________
## args[2] is input trajectory
#args[2] = "/mnt/databases/MD/PKM2/3BJT_monomer_apo/md0/3BJT_md0_ca.dcd"
message(paste("Input trajectory:", traj))
## trajectory name
traj_name = unlist(strsplit(traj, "\\.", perl = TRUE))[1]
## read structure
traj_bio3d = bio3d::read.dcd(traj)

#_______________________________________________________________________________
## args[3] is conformation increment if trajectory has been split
## the increment is added to the conformation number
message(paste("Output SA string numbering will start at #", confInc + 1))

#_______________________________________________________________________________
## xyz indices of Calpha atoms only, gleaned from PDB structure
ca.inds = bio3d::atom.select(str_bio3d, "calpha")

## assertion that highest CA atom index is within number of xyz coordinates 
stopifnot("Highest CA atom index outside trajectory atom indices" = range(ca.inds$xyz)[2] <= dim(traj_bio3d)[2])

#_______________________________________________________________________________
## encode trajectory
message("Encoding trajectory")
## trajectory is matrix of dimension 3N columns (xyz) and timesteps rows
sa_string.v = encode(traj_bio3d[ , ca.inds$xyz])
sa_name.v = as.character(c(1:length(sa_string.v)) + confInc) 

outFile = paste(str_name, traj_name, "sasta", sep = '.')
message("Writing SA-encoded trajectory to '", outFile, "'.")
## Structural Alphabet sequence in FASTA format
sa_string.fasta = sprintf(">%s%s%s%s%s\n%s",
                          str_name, ".", traj_name, "|", sa_name.v, sa_string.v)
## write SA-encoded trajectory as FASTA file
write.table(sa_string.fasta, file = outFile,
				quote = FALSE, row.names = FALSE, col.names = FALSE)

#===============================================================================



R/sa_encode.R
#===============================================================================
# DCDencode
#' sa_encode.R
#'
#' Encode a protein structure with the M32K25 structural alphabet.
#'
#' Given a PDB structure and trajectory, this function will fit C-alpha
#' coordinates of all consecutive 4-residue fragments with the canonical
#' fragments of the M32K25 structural alphabet, selecting the optimal
#' (lowest RMSD) fragment in each case.
#===============================================================================

library("pbapply")

#_______________________________________________________________________________
#' sadata: An S4 class for Structural Alphabet (SA) data.
#' @slot fragment_letters: Prototype fragment letters forming SA.
#' @slot fragment_coordinates: Coordinates of prototype fragments.
sadata <- setClass(
  "sadata",
  
  slots = c(
    fragment_letters = "vector",
    fragment_coordinates = "list"
  )
)

#_______________________________________________________________________________
## assign the M32K25 structural alphabet
.assign_M32K25 = function(sadata_o) {
  ## fragment letters
  sadata_o@fragment_letters = c('A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y');

  ## fragment coordinates: 4 Calpha atoms per fragment (vector)
  ## format per fragment: x1,y1,z1, x2,y2,z2, x3,y3,z3, x4,y4,z4
  sadata_o@fragment_coordinates = list(
    c(  2.630,  11.087,-12.054,  2.357,  13.026,-15.290,   1.365,  16.691,-15.389,   0.262,  18.241,-18.694), #A, 1
    c(  9.284,  15.264, 44.980, 12.933,  14.193, 44.880,  14.898,  12.077, 47.307,  18.502,  10.955, 47.619), #B, 2
    c( 25.311,  23.402, 33.999, 23.168,  25.490, 36.333,  23.449,  24.762, 40.062,  23.266,  27.976, 42.095), #C, 3
    c( 23.078,   3.265, -6.609, 21.369,   6.342, -4.176,  20.292,   6.283, -0.487,  17.232,   7.962,  1.027), #D, 4
    c( 72.856,  22.785, 26.895, 70.161,  25.403, 27.115,  70.776,  28.306, 29.539,  69.276,  31.709, 30.364), #E, 5
    c( 41.080,  47.709, 33.614, 39.271,  44.390, 33.864,  36.049,  44.118, 31.865,  32.984,  43.527, 34.064), #F, 6
    c( 59.399,  59.100, 40.375, 57.041,  57.165, 38.105,  54.802,  54.093, 38.498,  54.237,  51.873, 35.502), #G, 7
    c( -1.297,  14.123,  7.733,  1.518,  14.786,  5.230,   1.301,  17.718,  2.871,  -0.363,  16.930, -0.466), #H, 8
    c( 40.106,  24.098, 63.681, 40.195,  25.872, 60.382,  37.528,  27.160, 58.053,  37.489,  25.753, 54.503), #I, 9
    c( 25.589,   1.334, 11.216, 27.604,   1.905, 14.443,  30.853,  -0.042, 14.738,  30.051,  -1.020, 18.330), #J,10
    c( 17.239,  71.346, 65.430, 16.722,  74.180, 67.850,  18.184,  77.576, 67.092,  20.897,  77.030, 69.754), #K,11
    c( 82.032,  25.615,  4.316, 81.133,  23.686,  7.493,  83.903,  21.200,  8.341,  81.485,  19.142, 10.443), #L,12
    c( 28.972,  -1.893, -7.013, 28.574,  -5.153, -5.103,  30.790,  -7.852, -6.647,  30.144, -10.746, -4.275), #M,13
    c( -4.676,  72.183, 52.250, -2.345,  71.237, 55.105,   0.626,  71.396, 52.744,   1.491,  72.929, 49.374), #N,14
    c(  0.593,  -3.290,  6.669,  2.032,  -2.882,  3.163,   4.148,  -6.042,  3.493,   7.276,  -4.148,  2.496), #O,15
    c( 29.683,  47.318, 25.490, 26.781,  47.533, 27.949,  26.068,  51.138, 26.975,  27.539,  52.739, 30.088), #P,16
    c( 34.652,  36.550, 18.964, 33.617,  37.112, 15.311,  32.821,  40.823, 15.695,  34.062,  43.193, 12.979), #Q,17
    c(  8.082,  44.667, 15.947,  5.161,  46.576, 17.520,   5.855,  49.813, 15.603,   3.022,  50.724, 13.161), #R,18
    c( 64.114,  65.465, 28.862, 63.773,  68.407, 26.422,  67.481,  69.227, 26.232,  67.851,  68.149, 22.610), #S,19
    c( 18.708,-123.580,-46.136, 18.724,-126.113,-48.977,  18.606,-123.406,-51.661,  14.829,-123.053,-51.400), #T,20
    c( 61.732,  49.657, 35.675, 62.601,  46.569, 33.613,  65.943,  46.199, 35.408,  64.205,  46.488, 38.806), #U,21
    c( 88.350,  40.204, 52.963, 86.971,  39.540, 49.439,  85.732,  36.159, 50.328,  83.085,  37.796, 52.614), #V,22
    c( 23.791,  23.069,  3.102, 26.051,  22.698,  6.166,  23.278,  21.203,  8.349,  21.071,  19.248,  5.952), #W,23
    c(  1.199,   3.328, 36.383,  1.782,   3.032, 32.641,   1.158,   6.286, 30.903,   1.656,   8.424, 34.067), #X,24
    c( 33.001,  12.054,  8.400, 35.837,  11.159, 10.749,  38.009,  10.428,  7.736,  35.586,   7.969,  6.163)  #Y,25
  )
  names(sadata_o@fragment_coordinates) = sadata_o@fragment_letters;

  return(sadata_o);
}

#_______________________________________________________________________________
encode = function(traj.xyz) {
	sadata_o = new("sadata");
  
	sadata_o = .assign_M32K25(sadata_o);
	
	## N-3 structure fragments (and 3 coordinates per fragment)
	nConfFrag = (dim(traj.xyz)[2] / 3) - 3;
	nSAFrag = length(sadata_o@fragment_coordinates);
	
	## print progress bar also from RScript
	pbo = pboptions(type = "txt")
	## for each conformation (= row) in the trajectory (xyz matrix)
	sa_ali.v = pbapply::pbapply(traj.xyz, 1, function(traj_conf.xyz) {
	  ## fit all conformation fragments to all alphabet fragments
 	  ## for N-3 input structure fragments
 	  rmsd_m = sapply(1:nConfFrag, function(x) {
      	## and 25 SA fragments (structural alphabet letters)
		  sapply(1:nSAFrag, function(y) {
		  	#kabsch_R(traj_conf.xyz[((3 * x) - 2):((3 * x) + 9)], sadata_o@fragment_coordinates[[y]]);
		  	kabsch_C(4, traj_conf.xyz[((3 * x) - 2):((3 * x) + 9)], sadata_o@fragment_coordinates[[y]]);
	  	});
	  });

	  ## vector of minimal rmsd values
	  rmsd_min.v = apply(rmsd_m, 2, min);
	  ## vector of row indices of minimal rmsd values
	  rmsd_min.ix = sapply(1:length(rmsd_min.v), function(z) {
	   	which(rmsd_m[ , z] %in% rmsd_min.v[z]);
	  })

	  ## fragment string
	  sa_chars.v = sadata_o@fragment_letters[rmsd_min.ix];
	  ## from vector of chars to (vector of) string
	  sa_string.v = paste(sa_chars.v, collapse = '')

	  # return string of SA fragments
    return(sa_string.v);
	})
	return(sa_ali.v)
}

#===============================================================================
